File: Hard/41_First_Missing_Positive/Solution.cpp
#include "../../devlibs/cpp/cpphelper.h"

using namespace std;


// NOTE:
// For LeetCode submission, copy only the `class Solution` part.


class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        int target = 1;

        for (int num : nums) {
            if (num == target) {
                target++;
            } else if (num > target) {
                break;
            }
        }

        return target;
    }
};


int main(int argc, char* argv[]) {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    Solution solution;

    // test cases 1
    vector<int> nums1 = {1,2,0};
    cout << solution.firstMissingPositive(nums1) << endl;
    
    // test cases 2
    vector<int> nums2 = {3,4,-1,1};
    cout << solution.firstMissingPositive(nums2) << endl;

    // test cases 3
    vector<int> nums3 = {7,8,9,11,12};
    cout << solution.firstMissingPositive(nums3) << endl;

    return 0;
}

File: Hard/41_First_Missing_Positive/Solution.py
from typing import List, Optional


class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        nums.sort()
        target = 1

        for num in nums:
            if num == target:
                target += 1
            elif num > target:
                break

        return target


if __name__ == "__main__":
    solution = Solution()

    # test cases 1
    nums1 = [1,2,0]
    print(solution.firstMissingPositive(nums1))
    
    # test cases 2
    nums2 = [3,4,-1,1]
    print(solution.firstMissingPositive(nums2))
    
    # test cases 3
    nums3 = [7,8,9,11,12]
    print(solution.firstMissingPositive(nums3))